Fix schema serialization

khieta opened the pull request:
## Description of changes

Followup to #749 to fix json schema serialization.

* Added custom (de)serialization for `Id` and `Name`. This will affect our internal formats (e.g., used by DRT), but should not affect external formats, aside from error messages.
* This allowed me to remove some of the boilerplate in `schema_file_format.rs`

## Issue #, if available

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)

I confirm that this PR (choose one, and delete the other options):

- [X] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in cedar-spec, and how you have tested that your updates are correct.)

cdisselkoen commented on cedar-policy-core/src/ast/name.rs:
```diff
@@ -167,6 +178,34 @@ impl FromNormalizedStr for Name {
+    type Value = Name;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        formatter.write_str("a name consisting of a path and id, separated by `::`")
```
maybe this comment could make it clear that the path (and thus the `::`) is optional?  Not sure if that's easy to concisely convey though

khieta replied:
Updated to `"a name consisting of an optional namespace and id"`. Couldn't find a concise way to work in the `::` but open to suggestions.

cdisselkoen reviewed (approved):
Nice.  Shares a bunch of Deserialize logic that wasn't previously shared.  Also the error messages are improved

khieta commented:
Update: ran the `json-schema-roundtrip` target locally for a couple hours and all appears to be good.

john-h-kastner-aws commented on cedar-policy-core/src/ast/id.rs:
```diff
@@ -63,6 +63,34 @@ impl FromNormalizedStr for Id {
 
+struct IdVisitor;
+
+impl<'de> serde::de::Visitor<'de> for IdVisitor {
```
Add a comment that this impl is part of our public API now. Also for the other new impls used by json schema parsing.

As a side note, It'd be good if we could document more thoroughly how each `Serialize` and `Deserialize` implementation gets used. I've found my self wondering a few times if changing the internals of a struct would break part of our API, just DRT, or maybe nothing at all.

khieta replied:
I don't _think_ that any of the changes in this PR affect our public APIs.
* The `Id` and `Name` types are not publicly exposed, so changes to them should be safe.
* The only type with a `Deserialize` implementation in `cedar-policy` is `PolicyId`, which desugars to `SmolStr` (although possibly we should change that to be `Id`...).
* Most things in `cedar-policy/frontend` are represented directly as strings/JSON. The exceptions are `Decision`, `ValidationEnabled`, and `SchemaFragment`. The first two are not affected by this PR. The last is, but this PR fixes the deserialization to match what it was prior to #749

khieta replied:
Good point though on better documentation. Once we update DRT to use only public interfaces, we should be able to significantly cut down on the types that support (de)serialization.

john-h-kastner-aws replied:
My point is that this is `Deserialize` impl is now used to define the schema format (part of our api), but until now this `Deserialize` derive here only affected the DRT interface. It'd be easy to make a change here without thinking that it could break the json schema format

cdisselkoen replied:
> although possibly we should change [PolicyId] to be Id

Today, we allow PolicyId's that are any string, not limited to Ids

john-h-kastner-aws commented on cedar-policy-validator/src/schema_file_format.rs:
```diff
@@ -1340,31 +1277,31 @@ mod strengthened_types {
         });
         let schema: Result<ActionEntityUID, _> = serde_json::from_value(src);
-        assert_matches!(schema, Err(err) if &err.to_string() == "invalid action type: unexpected token `::`");
+        assert_matches!(schema, Err(err) if &err.to_string() == "invalid name `::Action`: unexpected token `::`");
```
Would it be easy to make this `invalid action name ...`?

cdisselkoen replied:
while we're here, this assert might be more pretty/readable written without the guard:
```
assert_matches!(schema, Err(err) => {
    assert_eq!(err.to_string, "invalid action name: ...");
});
```

khieta replied:
> Would it be easy to make this invalid action name ...?

I think we'd need to add back the manual deserialization function for `ActionEntityUID` to get a customized error message here. I don't think it's worth it, but open to other opinions.

john-h-kastner-aws replied:
Seems alright as it is

john-h-kastner-aws reviewed (approved):
Looks good. Just would like some comments on deser impls

khieta commented:
Another update: this PR breaks the Lean JSON parsing. Looks like CI didn't fail because we're not running the DRT implementation on the integration tests. I'll work on finishing off #488 before merging this PR.

khieta commented:
Ready for re-review.
* Added comments in `id.rs` and `name.rs` that de/serialization is used for the JSON schema format
* Updated to use the new downstream CI for `cedar-spec` (which now fails, as expected)
* Cleaned up some of the testing code

I'll now start working on the update to `cedar-spec` for this PR.

john-h-kastner-aws commented on cedar-policy-validator/src/schema_file_format.rs:
```diff
@@ -1406,38 +1351,174 @@ mod strengthened_types {
+mod test_json_roundtrip {
+    use super::*;
+
+    fn roundtrip(schema: SchemaFragment) {
```
```suggestion
    #[track_caller]
    fn roundtrip(schema: SchemaFragment) {
```

john-h-kastner-aws commented on cedar-policy-validator/src/schema_file_format.rs:
```diff
@@ -1111,6 +1047,11 @@ mod strengthened_types {
     };
 
+    /// Assert that `result` is an `Err`, and the error message matches `msg`
+    fn assert_error_matches<T: std::fmt::Debug>(result: Result<T, serde_json::Error>, msg: &str) {
```
```suggestion
    #[track_caller]
    fn assert_error_matches<T: std::fmt::Debug>(result: Result<T, serde_json::Error>, msg: &str) {
```
